### ISSUE TITLE
any ideas about the microcontroller or protocol the tag use?

### ISSUE THREAD

**Issue by mofosyne** [sylvek/itracing2#5]
I did a bit of a teardown of the device, but could not find the microcontroller data sheet it uses. How did you work out its communication?

[Btw here is the teardown of the iTag PCB](http://briankhuu.com/blog/self/hardware/2015/10/27/cheap-ebay-ble-tag-teardown-itag-via-itracing-app.html) . I find no mention of "ST17H25 datasheet" in google for the BLE SoC

**Comment by sylvek** [comment]
Hi,
the only information that i have is the name "Quintic PROXR".
your link is very interesting.

**Comment by mofosyne** [comment]
Your reference BLE tag you used is same in shape as mine?

**Comment by sylvek** [comment]
yes exactly the same… i suppose that the problem is under the android code of your phone :-/ :-/

**Comment by mofosyne** [comment]
True true. Btw what's your guess about "SWS" pin on the PCB? I deduced PWM1 as the LED, and BZ+ to be for the buzzer.

But I am not sure what SWS stands for. I wonder if its a way to "program" the microcontroller? Or is it serial output? Then again... there is also a possibility that the IC is hardcoded and locked.

**Comment by hosek** [comment]
Opened mine (http://i.ebayimg.com/images/g/NUQAAOSw3ydV5nNb/s-l500.jpg) same shape, same pcb, but the chip has different labeling:
TL SR8266
F512ET32
CK1526
cfapom 1p

**Comment by sylvek** [comment]
My device contains..  Vcc,  God, RX, TX, atm?, in, clk

**Comment by sylvek** [comment]
![20151027_213910](https://cloud.githubusercontent.com/assets/2259101/10771902/b6380898-7cf3-11e5-8ea7-2e3cf3fa93cd.jpg)
![Uploading 20151027_213208.jpg…]()
![Uploading 20151027_213900.jpg…]()

**Comment by sylvek** [comment]
![20151027_213910](https://cloud.githubusercontent.com/assets/2259101/10771929/c9f5d860-7cf3-11e5-8e95-9b8ef5097fbc.jpg)
![20151027_213900](https://cloud.githubusercontent.com/assets/2259101/10771932/ce204588-7cf3-11e5-9be8-934c893b2a07.jpg)

**Comment by sylvek** [comment]
KLJ-1230 for the buzzer
Q902t
434NJ

**Comment by mofosyne** [comment]
Man... You got a really beautiful board. It even has rx and tx. Which is
probably the serial port.

**Comment by mofosyne** [comment]
hmmm... do you think we should document this in the wiki perhaps?

**Comment by mofosyne** [comment]
Oh and btw sylvek, is the packaging for your tag the same as mine? Or did it look different?

**Comment by sylvek** [comment]
>Man... You got a really beautiful board. It even has rx and tx. Which is
probably the serial port.

you're the first guy to tell me that ^^

**Comment by sylvek** [comment]
>Oh and btw sylvek, is the packaging for your tag the same as mine? Is the name of the tag the same? Or did it look different?
my packaging was exactly the same !

**Comment by sylvek** [comment]
>Could try soldering to GND, RX, and TX to a serial converter and see what it says

i could, i have the serial 3.3V controller and an arduino if necessary

**Comment by mofosyne** [comment]
yea that would be interesting to see

**Comment by sylvek** [comment]
;)

**Comment by hosek** [comment]
Today is "singles day" on chienese markets and those tags are ~2$ and itag ~5$ so expect new wave of new requests or in worst case incopatibilities:/ 
I will add some photos of pcb when the second batch arrives

**Comment by sylvek** [comment]
i tried yesterday to weld some wire to the tx-rx "pin" … impossible. I'll try it an another day.
do you have a link to by some itag? (with paypal support if possible)

**Comment by mofosyne** [comment]
![brace-yourself-winter](https://cloud.githubusercontent.com/assets/827793/11090755/bc945fe4-88c8-11e5-8c33-a18a7a023458.jpg)

Have you considered asking these suppliers for images of the PCB?

**Comment by mofosyne** [comment]
Update: Just recently got another tag of the same shape, but from a different supplier that looked like it had a different PCB inside... it came... but with the same crappy PCB that I disassembled before 0.o

I have a feeling they got to the volume that they just opted for a factory preprogrammed or perhaps even an ASIC based chip (which is only economical in large volumes). Or just any methods that would allow them to do away with a programming pin pads.

**Comment by sylvek** [comment]
do you have a link to buy some iTag? i'm looking for it (needed to develop a version with support of several devices)

**Comment by mofosyne** [comment]
it was brought from ebay.com.au from a random selection of buyers. But I suspect all these buyers are buying from the same factory. e.g. alibaba or something

**Comment by hosek** [comment]
Mine pcb
![img_20151120_103912](https://cloud.githubusercontent.com/assets/9611327/11298003/5ae1b0e8-8f7c-11e5-8048-2506da4e406c.jpg)
![img_20151120_103921](https://cloud.githubusercontent.com/assets/9611327/11298010/5ec8bf62-8f7c-11e5-8743-c89b286c05a0.jpg)

**Comment by mofosyne** [comment]
```
TLSR8266
F5123T32
CK1528
CFAPCm 1P
```

 ? Looks to have same pinout as the `ST17H25` in the 3 separate tags I opened. 

```
1. First teardown mcu
    ST17H25
    F512ET32
    CK1528
    CFAPOJ 1P

2. Second tag from same store
    ST17H25
    F412ET32
    CK1528
    CFAPOL 1P

3. Third tag from different store
    ST17H25
    F512ET32
    CK1528
    CFAR2W 1P
```

---------


Now that I compaired all these tags. The `CK1528` seems to be pretty constant. But google search turns up nothing...

again I am reminded of this:

One broken url at eliabieri.com/65 :

```
my attempt on reverse engineering the cheapest bluetooth ...
eliabieri.com/65
Aug 19, 2015 - Inside I found the TLSR8266 which is produced by TELINK semiconductor. Besides that, theres a small buzzer inside, which can be used to ...
```

and this PDF in http://www.docin.com/p-878724807.html that I cannot download for some reason.

**Comment by mofosyne** [comment]
Does anyone speak chinese here? Can you download the pdf here? http://www.docin.com/p-878724807.html

Anyway these are what sticks out to me:

Pinout at page 71

<img width="504" alt="capture" src="https://cloud.githubusercontent.com/assets/827793/11300184/322c4004-8fe1-11e5-9f4e-c9c8b4aff080.PNG">

...

**!!!!!! SWS and SWM !!!!! at page 39 !!!!! This looks very very interesting   !!!!**

Ohhh.... so it means.... Single Wire Master and Single Wire Slave

<img width="415" alt="capture" src="https://cloud.githubusercontent.com/assets/827793/11300280/d2cfa294-8fe1-11e5-8d32-979f62a83e65.PNG">

![image](https://cloud.githubusercontent.com/assets/827793/11300317/200353b2-8fe2-11e5-933a-b9a6f560f6ac.png)

![image](https://cloud.githubusercontent.com/assets/827793/11300326/35df8912-8fe2-11e5-8f69-503ad8491217.png)

**Comment by sylvek** [comment]
wow great!
![20151120_140538](https://cloud.githubusercontent.com/assets/2259101/11300567/7934777e-8f90-11e5-9e7e-b5d3a434e05a.jpg)

mine use a Quantic chip (http://electronics360.globalspec.com/article/4758/nxp-buys-wearable-bluetooth-business)

ref: Q9021 434NJ OOUME

**Comment by zoranx** [comment]
[Telink_TLSR8266F512-datasheet.pdf](https://github.com/sylvek/itracing2/files/46678/Telink_TLSR8266F512-datasheet.pdf)

**Comment by hosek** [comment]
Got another batch of tags drom different seller (comes without batteries..great:/)
and they have the same pcb as I posted above, so this is probably "most general" type

**Comment by mofosyne** [comment]
You have my thanks "@zoranx". The next step is to verify if they locked the firmware rom down, and maybe to dump the firmware or something. 

But yea I agree, at a certain point, they usually settle to one pcb board "@hosek". The same thing happened with those flicker led lights. Started as a music chip hooked directly to an LED, then later on the chip was embedded into the actual LED itself.

**Comment by stanislav-brabec** [comment]
The original name of the chip, the price, list features and note about 32-bit RISC CPU (without mentioning type), indicates that this chip could be just a modification of ESP8266 that is programmed for BLE radio instead of WiFi radio.

If it is true, it could be programmed using SDK for Xtensa CPU, and toolkit for ESP8266. Just radio images would be different.

**Comment by sylvek** [comment]
that's sound good :)

**Comment by luciocorrea** [comment]
@sylvek based on what you said about "Quintic PROXR", I found this: http://en.t-firefly.com/en/firesmart/fireble/download/

It looks like there is an SDK and a git repository for QN9020 BLE chip related development. 

Is that useful?

**Comment by sylvek** [comment]
thank you @luciocorrea !

**Comment by sylvek** [comment]
you're right (http://wiki.t-firefly.com/index.php/FireBLE/Starter_guide/en)

**Comment by sylvek** [comment]
- SoC – NXP QN9021 ARM Cortex M0 MCU @ 32MHz with 94KB ROM (protocol stack), 64 KB SRAM, 128KB flash
- Bluetooth – BT 4.0 single mode. Central and peripheral mode with up to 8 simultaneous connections.

#awesome !

**Comment by sylvek** [comment]
instructive http://en.t-firefly.com/en/firesmart/fireble/
my own version uses this chip

**Comment by pfalcon** [comment]
Ok, so now the question is where to reliably buy a tag with QN9021 chip. @sylvek, so far, you seem to be the *only* owner of a tag with QN9021. Everyone else got TLSR8266 (some have obfuscated/rebranded markings). The 2 I got are also based on TLSR8266, and the second one I ordered because product page had a different PCB showing thru battery hole, but I got the same TLSR8266.

So, @sylvek, again, please let us know of reliable source to get QN9021-based ones. If you don't know of such, then let me summarize it like: the version you got was a branded one, or an early chinese clone. Shanzhai have now optimized their process and in all crappy cheap tags use crappy cheap chinese chips, so you won't be able to find anything else.

**Comment by sylvek** [comment]
@pfalcon i'm asking me the same… in fact, my itag is a present. This afternoon i've eaten with my friend who offered me this tag. He bought it on a website that sells different stuffs every days. So it's rellay tricky.. but the package seems to be this one (http://themouse.org/2014/12/19/test-de-themouse-itag-isee-genial-si-vous-perdez-tout-ne-cherchez-plus-vos-cles/)

**Comment by sylvek** [comment]
found that too: http://guru.multimedia.cx/bluetooth-tracking-devicestagskey-finders/

**Comment by sylvek** [comment]
if the picture is correct the back of this devices looks like mine
http://fr.aliexpress.com/item/5pcs-S1123-iTag-Anti-lost-alarm-Theft-Device-Anti-lost-Self-portrait-for-bluetooth-4-0/32295991777.html?ws_ab_test=searchweb201556_1%2Csearchweb201644_1_10001_10002_10005_10006_10003_10004_62%2Csearchweb201560_1%2Csearchweb1451318400_6148&spm=2114.06010108.3.223.aZgU2E

**Comment by pfalcon** [comment]
@sylvek : Thanks, here ~ the same in single quantity: http://www.aliexpress.com/item/Wireless-iTag-Self-Portrait-Anti-lost-alarm-Theft-Device-for-bluetooth-4-0-Smartphone-Support-iPhone/32296968392.html

But: last one I ordered was this: http://www.aliexpress.com/item/Hot-Smart-Bluetooth-4-0-Tracer-GPS-Locator-Tag-Alarm-Wallet-Key-Pet-Dog-Tracker-Hot/32438953423.html . Spot that it shows PCB which doesn't yet seem to have appeared on this thread, and there're even few GPIO signals broken out ;-). But, ordering that one, I instead got TLSR8266 one, with usual white silkscreen in battery compartment. But thanks to your another recent link - http://guru.multimedia.cx/bluetooth-tracking-devicestagskey-finders/ - we now know what SoC this [recent one](http://www.aliexpress.com/item/Hot-Smart-Bluetooth-4-0-Tracer-GPS-Locator-Tag-Alarm-Wallet-Key-Pet-Dog-Tracker-Hot/32438953423.html) has: BK3431 - the PCB shown in blog post shows similar GPIOs broken out, etc.. At least it's ARM, but not even a Cortex-M, it's ARM9.

So, all in all, with Chinese sellers, you cannot be sure that you'll get exactly what's on the picture. The only hint is price - with TLSR8266-based ones went below $2, the one in your and my link are $5-6.

**Comment by sylvek** [comment]
I'm trying to contact French resellers (who sell the same packages like mine)

**Comment by sylvek** [comment]
Nice touch... A French reseller took a picture and that seems to be a good one. (but more expensive, around 11€)

**Comment by jacksonliam** [comment]
Here's my board
![dsc_0037](https://cloud.githubusercontent.com/assets/3720701/12218996/fda1f5c6-b728-11e5-97a5-54da7dbdc8bb.JPG)
![dsc_0036](https://cloud.githubusercontent.com/assets/3720701/12218997/fda533ee-b728-11e5-9bd1-c8c7e40291b4.JPG)

st17h25 f512et32 ck1544

Does anyone know if the 'ST' chips are the same as the telink chips? Could be a clone? It looks like they're pin compatible/using the same PCB anyway. 

I'll order some more from random sellers (as they're under 2 dollars delivered!) and see what chips I get. 

It's a lot of BOM for under $2 if you ask me! 5 part Injection moulded case, coin cell, proper button, radio SoC, crystal, fet, buzzer, diode and a few passives! 

I can't see them hitting that price with the QN9020, it's $1.20 in volume.

**Comment by sylvek** [comment]
http://m.ebay.fr/itm/iTag-Anti-perte-bluetooth-Universel-cle-mobile-valises-voiture-animaux-/262228173326?nav=SEARCH

**Comment by verbage** [comment]
I just got a new tag with an ST17H26 chip, which is probably a slight revision to the ST17H25 chip folks have already reported.  It identifies as an MLE-15 vs. the more common iTag style.  It seems to work well with iTracing2, whereas the various iTag-version ones I have are problematic (yes, I realize sylvek said this is for MLE-15 tags...).  I'm not sure how to post images here, but I have pics of the device and guts if folks want to see it.

This came from Aliexpress ([http://www.aliexpress.com/item/Nut-Mini-Smart-Finder-Bluetooth-Tag-GPS-Tracker-Key-Wallet-Kids-Pet-Dog-Cat-Child-Bag/32579993483.html](url)), and it is currently US$3.23.  But identical ones are as cheap as US$2.64 ([http://www.aliexpress.com/item/Nut-Mini-Smart-Finder-Bluetooth-Tag-GPS-Tracker-Key-Wallet-Kids-Pet-Dog-Cat-Child-Bag/32559491726.html](url)).  There are several storefronts that also sell these, and though they might seem to be separate companies, they all seem to be related to some "YKS" distributor.  I have ordered a few more.

I also got a triangular one from the same company, and it has a Beken BK3431 chip inside, and identifies as an iTag device.  So I can't even connect to it with iTracing2.  Again, I realize sylvek said this is for MLE-15.  The URL to this triangular version is [http://www.aliexpress.com/item/New-Mini-Triangle-Smart-Tag-Wireless-Bluetooth-4-0-Tracker-Kid-Child-Bag-Wallet-Key-Pet/32579550627.html](url).

I do note that I never tried the original Chinese iTracing app, and came immediately to iTracing2.  Since I have only had iTag-identified devices up to now, things were just not working right with iTracing2.  But I just tried the original Chinese iTracing app, and they iTag-based devices work fine.  Yeah, that app is a bit quirky, and part of it is related to issues with English, but it seems to deal with iTag-version devices no problem.  By the way, it also works with my only MLE-15 device, too.

**Comment by sylvek** [comment]
@verbage Some users report to me that your triangular device will work too. That's the good news, the bad is that device is more bad ever.

**Comment by verbage** [comment]
@sylvek, curiously, with the BK3431-based triangular tag, even though it can be seen as an iTag device, I cannot seem to connect to it at all with iTracing2.  It just sits there with the spinning ball trying to connect.

I have a handful of teardrop-shaped, ST17H25-based devices (on an LC8266 PCB), and they also identify as iTag devices.  Though they are seen by iTracing2, it takes a couple of minutes to sync, and even then, when I try to have the phone set off the buzzer in the tag, sometimes it takes several to many times before it actually happens.  Sometimes it never happens.  There is an iTracing2 message--I forget exactly what it is--but something to the effect that a serious problem has occurred.  Again, I realize your work is for MLE-15 devices, but with the right documentation, problem supporting the iTag devices would be a possibility, too.

If you want a pic of the new ST17H26-based tag I have, let me know, but I am guessing this is probably just a slight revision of the ST17H25 chip.

**Comment by verbage** [comment]
And by the way, thanks so much for your efforts to support these cheap Chinese tags!

**Comment by jacksonliam** [comment]
@verbage can you post some pics of the front and back of the Belken board? 
You can't attach files in the mobile view so use the desktop view or a PC.

**Comment by verbage** [comment]
Here are pics of the front and back sides of the PCB.
![triang front](https://cloud.githubusercontent.com/assets/16998462/12718521/b1966354-c8b4-11e5-9362-dfe619df26bc.jpg)
![triang back](https://cloud.githubusercontent.com/assets/16998462/12718525/b68cae68-c8b4-11e5-8f7a-c7ed514bc150.jpg)

**Comment by jacksonliam** [comment]
Thanks, that's pretty cool. There's lots of test points on that PCB.

Can't find any sources for that chip in the usual places, but there's a module on taobao and an English datasheet http://www.datasheetspdf.com/PDF/BK3431/945631/6
It's an ARM core so should be easily programmable, but can't find any API/SDK.

**Comment by verbage** [comment]
Yes, it should be programmable unless they've done something to make it a one time burn.  I have a few more coming so I'm not averse to sacrificing this for the science to solder on leads and try stuff.

The battle here is know just which chip you'll get, but as I mentioned, the distributor/OEM seems to be this "YKS" company.  For example, in the link I pointed out for this triangular one, the Aliexpress store is called "Shenzhen CarNival Trading Co., Ltd.", but the label on the shipping package actually mentioned "szyks" (so maybe dropshipped?), and on the webpage, you can see under the product description that the bitmap headers have a "YKS" on the right.

Yeah, certainly "YKS" could start using a different stock of chips, but for the moment I am guessing I can count on ST17H26- and BK3431-based tags so I have ordered a few more.

**Comment by verbage** [comment]
Here is a pic of the Nut Mini PCB, which has the newer (I'm guessing...) ST17H26 chip.

![nut mini pcb](https://cloud.githubusercontent.com/assets/16998462/12735348/c1c0aa5c-c90c-11e5-8492-bc85bb5f8f9c.jpg)

**Comment by jacksonliam** [comment]
Wow I don't think I've seen a bluetooth chip that wasn't QFN or BGA before.
I bet that antenna doesn't work well at all. I don't think that PCB would be any cheaper to make than mine. Perhaps it's a little smaller, looks like it uses a different coin cell.

**Comment by verbage** [comment]
Interesting, I did not even think about the package type, but maybe easier to solder...  I can't really say anything about antenna performance because I have not done any systematic testing--just tested it nearby to make sure it worked.  I should have put a scale, but this uses a CR2032 battery so it's about the same size as other ones.

**Comment by sylvek** [comment]
Keyper - A Rechargeable Bluetooth Tracker Device, via @Kickstarter https://www.kickstarter.com/projects/keyper/keyper-a-rechargeable-bluetooth-tracker-device

**Comment by NicolaiJAndersen** [comment]
I am also the owner of a BL-108 tag, and it looks not very hackable.
I know that there are some other options for DIY BLE Tag, but probably a bit expensive.

If some have recommendations for a specific model (chip), please write it here.

**Comment by NicolaiJAndersen** [comment]
I have an idea for a tag, it must have a power On / Off button and a separate button for the function, the function button needs to be replaced with a vibration switch. (Alarm tag)

This requires that the tag, either always on (no power switch) or each input to power switch and vibration switch.

The easiest thing would of course be to modify a tag.

![1767-02](https://cloud.githubusercontent.com/assets/17329958/13170261/1f9c4ee2-d6eb-11e5-95e0-d0aa1bc372bf.jpg)

**Comment by iotopenproject** [comment]
> Keyper - A Rechargeable Bluetooth Tracker Device, via @Kickstarter https://www.kickstarter.com/projects/keyper/keyper-a-rechargeable-bluetooth-tracker-device

I think It is not good idea.
http://www.bluetoothcarkitsupplier.com/products/Key-finder.htm
http://www.aliexpress.com/store/product/HOOKE-HOK29-Smart-Intelligence-Tracer-Bluetooth-4-0-iTag-Bluetooth-Shutter-Finder-Anti-Lost-Keychain-Mother/1961310_32584766768.html

**Comment by stanislav-brabec** [comment]
@iotopenproject Thanks. I reported them here: http://kickscammed.com/ticket/keyper-a-rechargeable-bluetooth-tracker-device/

**Comment by iotopenproject** [comment]
Can anyone help? I need ST17H26-based tag circuit diagram.

**Comment by sylvek** [comment]
i've just received my new iTag (name iTag… and not QUANTIC PROXR like my previous one).
… and this inside box is the same than my previous one :+1: 

![20160308_185210](https://cloud.githubusercontent.com/assets/2259101/13611083/0b105c00-e561-11e5-9cf1-338923d2d4a9.jpg)
![20160308_185216](https://cloud.githubusercontent.com/assets/2259101/13611114/285e596a-e561-11e5-9b5e-c1ff89206e6d.jpg)
![20160308_185411](https://cloud.githubusercontent.com/assets/2259101/13611115/287a5e94-e561-11e5-996c-4972e1c1e3b9.jpg)
![20160308_185416](https://cloud.githubusercontent.com/assets/2259101/13611116/288235b0-e561-11e5-8dd5-891301e5daa3.jpg)

**Comment by olewales** [comment]
Hello, I've been doing some googling on telink chip and I think I've found something that may interest you: http://pan.baidu.com/s/1hqAfnc4
Additionally, since download speeds from baidu are appalling (for me at least) I did re-upload the whole  thing to more accessible service: https://mega.nz/#F!GZkVVAaQ!u0UmS6mOvPNJYV2Kyds3TQ

Contents seems promising at the first glance, but I have yet to do more thorough investigation. Let me know what you think

**Comment by projectgus** [comment]
Thanks for posting that @olewales, that's very interesting. I have a couple of Telit TLSR8266-based tags in a drawer somewhere, so I had a quick poke around out of interest. A few things I noticed:

* The Eclipse-based SDK installer failed for me on both a Windows 10 and Windows XP VM. I get an error part way about a file write error, and the install aborts. Maybe it needs Chinese Windows?!? It seems to fail before most of the interesting files have installed (Eclipse extracts, and a full mingw root, but nothing much else.)
* Some of the project source directories already have Makefiles in them (autogenerated by Eclipse), so you can poke around and see what the compile/link arguments are.
* The datasheet in the SDK files is (I think) the same datasheet that @sylvek posted above, and the only other docs seem to be explaining Telink's development/EVK boards.
* However, the source files for main MCU functions seem pretty complete. Plus there's a linker script with the memory map. The only binary blobs seem to be libraries for the BLE functionality, and they have headers.
* The TLSR8266 processor architecture seems to be "TC32" (the binary object files have ELF e_machine value 0x8800), and the compiler is gcc-based `tc32-elf-gcc` (which makes it a GPL violation for Telit to distribute their SDK without compiler source, not that I think they care.)
* There's an old precompiled "tc32" toolchain for Windows published here: https://github.com/boo996/t32eclipse The changelog file in the repo (from 2011) has the email address of Peter Shieh @ Telink.
* In April 2016 Peter Shieh also published the full source code for the Telink Eclipse plugin, here: https://github.com/TelinkDev/EclipseTC32Plugin and the README there suggests that the toolchain is coming to github too.

If anyone is really keen to work on this then contacting Peter Shieh directly might be a good idea (maybe Telink has seen the success of companies like Espressif providing publically downloadable SDKs, and want to open up their SDK as well.)

For my part, I'm probably done looking into this (I was hoping it might be an ARM core), but maybe something here will pique someone's interest. :)

**Comment by petiolarid** [comment]
Thanks sylvek,

I have three devices. All bought from Amazon but from two different sources. Interestingly the ‘Local Name’ of the devices are slightly different. My initial purchase has a ‘Local Name’ of ITAG (all capitals) whilst the other two (from a different source) are iTAG. Subtle but interesting.

One of the things I tried to do was to change these names (to make them unique). My write instructions returned without errors but when I read the names again they had not changed.

Of course I’ve tried replacing the battery on each of my devices – to no avail.

Today, my single working iTag has failed me. It was on my keyring for the last couple of days just accompanying me around the country and each time I looked at the iTag the little blue light was blinking. I made no attempt to interact with it during this time. This morning I noticed that the blue light had gone. No amount of button pressing or battery replacing has made it either beep or show a blue light. Also my scanning for LE devices finds nothing.

I’m guessing that the only way to ‘reset’ these devices is to throw them away and buy another!

Maybe I’ll try to momentarily put 3V onto the RST spot on the circuit board and see if that does the trick. I guess I have nothing to lose. But given that I’ve pretty much decided that these devices are at best, unreliable, this would hardly provide a workable solution even if it worked.

Perhaps I will order myself one of the ‘Nut’ devices – a bit more expensive but maybe they will work better for me.

**Comment by denysvitali** [comment]
Just wanted to give my contribute

I ordered an iTAG device for a personal project, and wanted to tear it  down to discover if that annoying sound can be removed (maybe it's easier to short out the buzzer).
I have tried to debug the TX / RX port via a JTAG without success

Here are some photos (don't judge my ability at soldering :smile:, I did it while I was really, really tired)
![img_1852](https://cloud.githubusercontent.com/assets/4939519/18256891/db34589e-73bb-11e6-845e-3ae6a790d88a.JPG)
![img_1835](https://cloud.githubusercontent.com/assets/4939519/18256903/f59d1f9a-73bb-11e6-87b0-38a9e10d0737.JPG)

**Comment by sylvek** [comment]
great! do you have traces ??

Le lun. 5 sept. 2016 à 22:57, Denys Vitali <notifications@github.com> a
écrit :

**Comment by denysvitali** [comment]
Unfortunately not, seems like dead, no data on TX / RX. I'm probably doing something wrong

**Comment by sylvek** [comment]
@denysvitali don't loose your faith.. you should set a another pin to 1 and force the reset (or something like that... without documentation, it's pretty hard) :(

**Comment by denysvitali** [comment]
@sylvek Sorry if I go off-topic, but apparently your app doesn't seems to detect the button press. I tried with nRF Connect and I can get the device services, subscribe to the "Key Press State" service and get the status when is pressed. Apparently in your app I cannot detect the key press state.

I'm trying now to write an app based on your source, but the button event is never fired :(

**Comment by sylvek** [comment]
strange... what it your device?

Le lun. 5 sept. 2016 à 23:15, Denys Vitali <notifications@github.com> a
écrit :

**Comment by denysvitali** [comment]
Nevermind, tried the beta version and it works. The device is a `iTAG`, the one I posted the pictures about.

Does the beta version have the same source code as this repo?

**Comment by sylvek** [comment]
strange because the current beta version IS the release version :]
the corresponding source code is contained into the branch "master".

Le lun. 5 sept. 2016 à 23:19, Denys Vitali <notifications@github.com> a
écrit :

**Comment by denysvitali** [comment]
Wonderful, tested again and it magically worked. Sorry for the OT and for wasting your time!

As soon as I have more free time I'll for sure deeply analyze this bluetooth button and report here the results

**Comment by sylvek** [comment]
you're welcome :]

Le lun. 5 sept. 2016 à 23:30, Denys Vitali <notifications@github.com> a
écrit :

**Comment by FSJO** [comment]
Bonjour sylvek. 
ive also become interested in the iTag dongle which advertises MLF-15. 

I've tried pinging different resellers on alibaba to get DoC documents (declaration of conformity) but without success (FCC & CE).  Have anyone here successful got a copy of it from the local seller?

What kind of problem have you seen with it? 
With your android app; It looks like the module crashing when disconnected / out of range, or perhaps something goes wrong with re-bonding ?  I still didnt look deep into to this thought I'll ask here first. 

Regarding the uart pins, perhaps the bootloader needs to be in a certain state for reflashing the module.
Scoping the pins could revile it during boot. I 'll look into it.

**Comment by sylvek** [comment]
hi @FSJO with the MLE-15, i'm facing a power off of the module when the connection is lost.

**Comment by SlowBro904** [comment]
@FSJO I'm a n00b to all this so sorry if this is a dumb question but have you looked for the DoC at the FCC search? I found docs on another module I was interested in. I did have to search for just the model number.

**Comment by FSJO** [comment]
I have no fcc id printed on the pcb, but yes otherwise it should be possible.
do you have a ID starting with 3-4 letters(manuf id) then a number?

**Comment by SlowBro904** [comment]
Not on mine. When I said I found docs it was on another module I was interested in.

**Comment by JendrikDerWeise** [comment]
![nutmini](https://cloud.githubusercontent.com/assets/9480085/20728851/7346292c-b634-11e6-8bb3-a8d1ae019ccb.JPG)

Hi, I have a nut mini and I´m wondering if it is possible to flash it. Does anybody know? And if yes, could you give me a hint how? The writing on the PCB is not really helpful, and I don´t know how to figure out what is what.
My goal is to connect a TAG to an Arduino... sorry for the offtopic, but I couldn´t finde really much informations...

**Comment by petiolarid** [comment]
I have a Nut Mini that I've used for a project with a Raspberry Pi. The Nut Mini does not have a button and is constantly broadcasting whilst the other models currently available (Nut 2 and Nut 3) do have a button and only broadcast a Bluetooth signal for a brief period when their 'button' is pressed.
For my purposes the Nut 2 and Nut 3 are much more useful.

**Comment by JendrikDerWeise** [comment]
I need a presence detection. So the nut minis broadcasting is exactly what I need. But I have no clue how to see it with Arduino...

**Comment by FSJO** [comment]
did you try using a raspberry with bluez or other bluetooth stack with BLE support? 
I think thats the easiest way instead of arduino.  However, make sure the scan completes otherwise the radio might hang.  If it gets unresponsive its a good idea to control the bluetooth module/usb-dongle with a GPIO thus being able to restart it. 

Do you have different mac_addresses or UUID on your device ?

**Comment by sylvek** [comment]
Indeed, you have to try an ip scanning for example.. When your phone is
connecting on your WiFi Hotspot.. Or write a same thing with a Bluetooth
dongle. It is not so hard.

Le mer. 7 déc. 2016 08:24, Fredrik Sjöholm <notifications@github.com> a
écrit :

**Comment by shyba** [comment]
Hi. http://www.lenzetech.com/product/showproduct.php?lang=cn&id=82
ST17H25 is a Lenzetech product.

Tip: when looking for this, always try to replace numbers with 'x'. Like: ST17Hxx

**Comment by shyba** [comment]
@mofosyne just got a very cheap ($1.85) bluetooth selfie shutter to tinker with and it came with ST17H26. It's like [this one](https://www.amazon.com/UFCIT-Bluetooth-Wireless-Control-Smartphones/dp/B00LUNAA2O/ref=sr_1_4?ie=UTF8&qid=1483767975&sr=8-4&keywords=remote+selfie+shutter), but totally unbranded.
Found it's datasheet on Docin. No idea how to download it, Google Translator is breaking the UI buttons. So I am reading on the website.

http://www.docin.com/p-1814297477.html

ST17H25 here: http://www.docin.com/p-1814343474.html (hey, this IC is way better than the one I got! Mine is OTP ROM)

It has the Single Wire Slave pin as well, but for my DIY project I think i2c or even the buttons pads are enough. I am trying to make a DIY cadence/speed sensor for my bike using this cheap selfie shutter + hall effect sensors, but for this price I think I will keep buying them for Bluetooth 4.0 DIY projects. :smile: 

Something I noticed on their datasheet: There are no scary confidential warnings or something like this, but mail addresses for sales and information! :+1: 
I think it should be fine to contact them if any of you is in the process of creating a commercial project and looking for a good and cheap BT SoC.

I hope it helps you.

**Comment by lbodnar** [comment]
This MLE-15 tag also nicely violates basics of BLE specification.

When button is pressed it continuously sends ADV_IND advertising packets every 17 milliseconds.
Specification calls for minimum 20ms plus random interval between 0ms and 10ms to allow two identical devices to not jam each other continuously.  So the repeat interval should be 25+-5ms.

Leo

**Comment by 0to1a** [comment]
I have buy much JDY-10 (they said it's CC2541) from China and the name of IC is blank. I saw the name is TLSR8266, how to program it?

**Comment by SlowBro904** [comment]
I have an idea for reprogramming these. Press some moldmaking silicone over the MCU, once it cures pull it off, and very carefully paint small traces using conductive paint out to some breakout points. Tada, you now have a reusable programming "hat." Buy dozens of these at $2 each and reprogram them for a longer life.

It's an idea I'm pursuing for a chicken tracking device.

**Comment by MD2020nMe** [comment]
There is a freeRTOS port of Telink chip that can possibly useful.  Especially the drivers.   This works well on their 8269 (pin to pin to 8267 with larger SRAM).   JW played with the OpenOCD for this chip.   He can possibly be the source of the tools.  There are some tools floating around.   JW may have them.

**Comment by dwaynez** [comment]
My itag with the ST17H26 chip in it does not seem to advertise the link loss service. Without this service, it appears that there is no way to tell the tag not to alert on disconnect from the phone. Is anyone aware of a cheap tag that provides a link loss service that the tag could be told to not signal on link loss?

**Comment by mariarti** [comment]
more geek porn
![bluetooth_face](https://user-images.githubusercontent.com/1933140/27486128-c9fc21ca-582f-11e7-84a0-ac5ed68a90ca.jpg)

**Comment by brailateo** [comment]
Hello all! I recently bought 2 devices with ST17H26 chip (CH1615) and I'm using them as a "Panic Button" for my old parents. It seems to be the perfect "use case", they are off all the time (no battery use) and when needed, long press put the tag ON, I have 4 old phones and tablets (Android 4.4.2) all over the place in house with a small program that detects them.
Everything is ok, BUT ... the range seems to be too small (approx. 3 or 4 meters). Did anyone tried to replace the MIFA PCB antenna with a Wire Antenna (lambda/4) in order to increase the range?

Thanks for itracing2, it helped me as an inspiration source for my detection program,
Teo
![antenable](https://user-images.githubusercontent.com/602966/28590501-ff8dc78e-718a-11e7-94f4-97cec68118c3.jpg)

**Comment by barsznica** [comment]
Looks like a neat little IC, already packaged. I'm interested in doing the same, but where is this "relevant file" referred to below? "128K bytes FLASH memory is used for storing program code. It can be accessed through a simple SPI interface. Four GPIO P0.4, P0.5, P0.6 and P0.7 are used for program FLASH in program mode. You can read, write or erase the FLASH space with different SPI command. For detail information, please refer to the relevant file. After erase, all the data in the FLASH are 0xFF, customer can write their program into the FLASH. Also, the data in the FLASH can be read out for debug. Once the developing process is done, the access for the FLASH can be forbbided by writing 0x12345678 into the last four addresses. ([128k-4]~[128k])", https://www.google.com.au/url?sa=t&rct=j&q=&esrc=s&source=web&cd=11&ved=0ahUKEwjWgu788LrWAhXEWbwKHYxmAeg4ChAWCDYwAA&url=http%3A%2F%2Fwww.9mcu.com%2F9mcubbs%2Fforum.php%3Fmod%3Dattachment%26aid%3DNjA1NDZ8NjZmOGNlMDl8MTQ4NDkwMjY5NXwwfDEwNzg3NzA%253D&usg=AFQjCNEOo0hViuf0z3YOueaEoEQt3eADiA, so maybe it has been "forbidded"?

From the datasheet: FLASH programming: JTAG, I2C, SPI and UART interface. It looks like P04, 5, 6 & 7 are connected to the SPI (4-MOSI, 5-MISO, 6-SCK & 7-CSI (CE)), while the UART interface is also broken out. PROG sets "Flash Program Mode selection; 0: Flash in normal mode & 1: Flash in program mode"???

I think this a dead end, but it was fun trying...

![imag0907](https://user-images.githubusercontent.com/11470190/30773161-07c95b2c-a0aa-11e7-934f-de1700897d94.jpg)
![imag0908](https://user-images.githubusercontent.com/11470190/30773164-0db969aa-a0aa-11e7-9624-0f31ccedd79f.jpg)

**Comment by jpmorrison** [comment]
Found a download link to the Nanosic SDK which appears to have the compiler for Telink MCUs - tc32-elf-gcc. No source code unfortunately and emails don't work to nanosic.com

Forked:
https://github.com/jpmorrison/NANOSIC_IDE.git

**Comment by zerog2k** [comment]
regarding the beken bk3431n - I found this random keil project in github... unsure of how useful it might be to start a blinky/helloworld with a jlink... (Not sure if the flash lock is permanent or resettable):
https://github.com/SuperCoderMan/ble/tree/3d237f8e37c3a989d62c87349219e7691868b971/BK3431_designkit_V2.7

**Comment by fanoush** [comment]
Hello, anybody got further with flashing Telink TLSR8266/ST17H25 chip via SWS? I got few nice cheap fitness trackers based on this chip, the SDK with gcc appears to work, I have matching firmware binaries for those bracelets as a backup, but it looks like one would currently need the "Telink EVK OTP/Flash Programmer board" to reflash it. Sadly I can't google this programmer board anywhere being available (aliexpress, ebay).

The protocol is poorly documented and only in older versions of the datasheet, later this SWS/SWM part was removed completely. Still, I checked the protocol description and while the basics are good I can't figure our what should be the clock speed for data (unlike with CC2541 where there is extra wire for clock). Also there is only read and write command documented (and begin, end) , how would one actually execute something in the chip via this protocol? How could the tcdb debugger be implemented with this? Maybe backing up and rewriting stack or ISR vectors? Or maybe CPU registers including PC and stack pointers  are memory mapped? Would be nice to have some USB debug traces of this EVK programming board doing something.

**Comment by tablatronix** [comment]
For posterity, I just got these 2, seem to be different revisions

![28096eac-2aaf-4dd8-80db-43a6edb5ff62](https://user-images.githubusercontent.com/807787/47055836-26fd8200-d17f-11e8-93cf-24466eb24f10.jpeg)
![e843676c-c4d3-4c25-b16f-b0f62d99a132](https://user-images.githubusercontent.com/807787/47055837-26fd8200-d17f-11e8-803b-bdc3eb3a1e24.jpeg)
![c26a827f-ee32-4864-895f-26d9514f7f7d](https://user-images.githubusercontent.com/807787/47055838-26fd8200-d17f-11e8-881b-e06cb7d68600.jpeg)
![80774043-3751-43b0-afb6-0f959e74c372](https://user-images.githubusercontent.com/807787/47055839-27961880-d17f-11e8-8201-b27711fa463c.jpeg)

**Comment by feversky** [comment]
> Hello, anybody got further with flashing Telink TLSR8266/ST17H25 chip via SWS? I got few nice cheap fitness trackers based on this chip, the SDK with gcc appears to work, I have matching firmware binaries for those bracelets as a backup, but it looks like one would currently need the "Telink EVK OTP/Flash Programmer board" to reflash it. Sadly I can't google this programmer board anywhere being available (aliexpress, ebay).
> 
> The protocol is poorly documented and only in older versions of the datasheet, later this SWS/SWM part was removed completely. Still, I checked the protocol description and while the basics are good I can't figure our what should be the clock speed for data (unlike with CC2541 where there is extra wire for clock). Also there is only read and write command documented (and begin, end) , how would one actually execute something in the chip via this protocol? How could the tcdb debugger be implemented with this? Maybe backing up and rewriting stack or ISR vectors? Or maybe CPU registers including PC and stack pointers are memory mapped? Would be nice to have some USB debug traces of this EVK programming board doing something.

hello fanoush. recently i found the company repository https://github.com/17HXX, hope it helps. the EVK board is only available in the company official account in Wechat, but very expensive, 1000RMB. anyone who needs translation or other information, please contact me. i had the idea to make a firmware, and use it as a positioning device for home automation.  maybe replacing with a cc2540 is much eaiser.

**Comment by fanoush** [comment]
> 
> hello fanoush. recently i found the company repository https://github.com/17HXX, hope it helps. the EVK board is only available in the company official account in Wechat, but very expensive, 1000RMB. anyone who needs translation or other information, please contact me. i had the idea to make a firmware, and use it as a positioning device for home automation. maybe replacing with a cc2540 is much eaiser.

Thank you very much, I already searched github repositories for 'tlsr' and 'telink' but forgot 'lenze', however when trying now it wouldn't find this 17HXX anyway so thanks a lot for this pointer! cc254x is 8051, which does not look easier to me, especially with no free compiler, using IAR 30 day trials gets boring very quickly :-) I managed to recompile BlueBasic for it and tested it with JDY-08 module and it is quite nice  but both 8051 arch and 8k  SRAM just feels too limiting. I'll definitely check stuff in 17HXX out and see whether it helps me with figuring out how to bitbang SWS. I already got JDY-10 TLSR8266 module for this.

**Comment by code0100fun** [comment]
Sorry if I'm reviving an old discussion but I came across this chip tearing down a "CSI TECH" beacon and I found a datasheet in English that may be useful for some of you https://share.weiyun.com/5JawPYc

**Comment by fanoush** [comment]
As for the TLSR8266 I am not sure if I missed it previously or they added it later but there is link to wiki http://wiki.telink-semi.cn/dokuwiki/doku.php?id=start directly on main Telink page. It contains datasheets, IDE with compiler, BLE SDK. Also there is documentation about Burning and Debugging Tool (BDT) with downloads containing its schematics and firmware binary. It still does not help me to know more about SWS but at least it is lot of (mostly known) info in one official place. BTW from schematics it looks like BDT itself contains TLSR8266 chip which can natively handle USB - the chip has pins for usb data +/- However the TLSR8266 datasheet has no description of USB functionality.

**Comment by PlantDaddy** [comment]
Finally found the right datasheet for the ST17H26:
https://datasheetspdf.com/datasheet/ST17H26.html

PDF page 23, page 22 in the document
https://datasheetspdf.com/pdf-file/1316228/LENZE/ST17H26/1

**Comment by fanoush** [comment]
@PlantDaddy not sure what you're after but some version of that datasheet is also in github project mentined above, see https://github.com/17HXX/17HXX_OTP_Plug-in_Flash_SDK_Release/tree/master/doc , there may be more information there

**Comment by probonopd** [comment]
OTP = one-time-programmable = the end of all firmware hacking dreams?

**Comment by fanoush** [comment]
17H26 has OTP yes, 17H25 has flash like TLSR8266

**Comment by arteyshock** [comment]
Maybe something like this https://ru.aliexpress.com/item/33045005525.html?spm=a2g0s.8937460.0.0.7b032e0eC7oFuA could be used to program TLSR8266

**Comment by ther12k** [comment]
any idea to make this device to always on?

**Comment by pvvx** [comment]
Telink EVK on the E104-BT10-G/N module (TLSR8269).
https://github.com/pvvx/TLSR8269-EVK

https://aliexpress.com/item/4000032659038.html
https://aliexpress.com/item/4000032948828.html

**Comment by fanoush** [comment]
So how do you write the EVK firmware to new module without having EVK? Over USB? And then it behaves over USB as EVK? So is there some usb bootloader/flashing mode so you can write different firmware to it even when it is already flashed with EVK (or any other) firmware?

An maybe the same procedure could even work with 8266 chip too over USB? There are some USB pins on 8266 too. And the real EVK has 8266 inside too.

**Comment by pvvx** [comment]
USB only works with software support.
ISP is not in 8269. This is only advertising.
There are many examples of working with USB in the Telink SDK

**Comment by pvvx** [comment]
Alternative programmer for TLSR826x on STM32F103C8T6 beta version
https://github.com/pvvx/TlsrTools

**Comment by pvvx** [comment]
To fanoush: 
E104-BT10-G/N comes with firmware. It has software support for reprogramming via USB. It is also possible to program update the chip via BLE in an Android application.

**Comment by pvvx** [comment]
TLSR8266 (AT32) - no SWM pin. TLSR8269 - there is a contact SWM (PA7). Both have USB.

**Comment by bunnypcb** [comment]
> 
> 
> any idea to make this device to always on?

You need to modify the firmware in order to get rid off the delay to turn on or to enter in the stand by.

**Comment by bartowl** [comment]
for some of you still interested in ST17H25, here is an english datasheet:
https://datasheet.lcsc.com/szlcsc/1811151231_LENZE-ST17H25_C326546.pdf

**Comment by Bresenham** [comment]
Where can I buy the programmer "Burning EVK" from Telink?

**Comment by atc1441** [comment]
Hey. So i was able to flash and read out the TLSR8266 and TLSR8251 micros with a genuine Telink flasher
Tested on some cheap 2,5€ fitness Tracker and the 4€ Mi Thermostat
https://twitter.com/atc1441/status/1294359879696764928?s=19

Also here is a small demo of it https://youtu.be/md-CoIWJ3kc

The SDK i am using now is this one:
https://github.com/Ai-Thinker-Open/Telink_825X_SDK

On some TLSR versions its needed to use the SWS pin and on other the SWM pin

**Comment by Bresenham** [comment]
Do you mind posting a link where you bought the "genuine Telink flasher"?

**Comment by exikyut** [comment]
Someone asked the same question on YouTube, and was pointed to a sadly now-dead Aliexpress link.

It was still in Google's cache though, and following the link to the seller store found this (USD$72): https://www.aliexpress.com/item/33045005525.html

The SWS protocol has been figured out though, https://github.com/pvvx/TlsrComSwireWriter bitbangs SWS over a standard USB UART adapter to run programs from memory, and https://github.com/pvvx/TlsrComProg extends this to allow flashing, with the prerequisite that you have the TX and RX pins available (presumably these allow faster I/O than the SWS pin does).

<sub>Sadly the thing I'm trying to tweak doesn't expose RX, yay.</sub>

**Comment by atc1441** [comment]
Was able to MODIFY the uart flasher to directly flash the TLSR8266 and TLSR8251. 

That way the genuine flasher is not needed anymore at all. 

https://github.com/atc1441/ATC_MiThermometer

The @pvvx version does need a bootloader to be flashed first

Also got the Telink OTA update protocol running in a web flasher, that way its possible to update the devices who have the telink ota characteristic

**Comment by atc1441** [comment]
Here is also a video about that uart flasher

https://youtu.be/mt-9jcCwmFA

**Comment by emma107946** [comment]
If you are a serious gamer and playing in front of a computer screen, here you can buy your favorite gaming glasses and prescription glasses that protect your eyes from blue light.
[Prescription Glasses](https://www.syghtglass.com/collections/prescription-gaming-glasses)
[what do gaming glasses do](https://www.syghtglass.com/blogs/gaming-glasses/do-gaming-glasses-really-protect-your-eyesight)
[computer glasses blue light](https://www.syghtglass.com/collections/computer-glasses)
[yellow gaming glasses](https://www.syghtglass.com/pages/crystal-clear-vs-amber-tinted-lens)
[Clip on Glasses](https://www.syghtglass.com/collections/gaming-glasses-clip-on)
[night vision glasses for driving](https://www.syghtglass.com/products/night-driving-glasses)

**Comment by yeshapatel1210** [comment]
A [**white gold tennis bracelet**](https://www.shantijewel.com/natural-diamond-bracelets/tennis-bracelet.html) is an iconic jewelry item that is suitable for all occasions. Buy diamond bracelets or bracelets online from us at the best price. We have tennis bracelets set in yellow, white, rose gold, or platinum.
